File: apps/gateway/src/governance/permissions.ts
/**
 * Permission mode framework — faithful Node port of
 * packages/superclaw/src/superclaw/permissions.py.
 *
 * Design rule (unchanged from Python):
 * - SuperClaw exposes exactly TWO presets to users: `ask` and `allow`.
 * - SuperClaw makes NO per-action decisions of its own — it is a pass-through
 *   that translates a preset into whatever the underlying runtime supports.
 * - Both presets map to the runtime's MAXIMUM ("bypassPermissions"). `ask` must
 *   NOT claim to prompt/restrict — that would be surface fraud (permissions.py:41).
 * - No decision engine, no action taxonomy, no third preset. Those were
 *   considered and rejected as over-engineering (permission-broker-plan.md).
 *
 * This module deliberately stays tiny and behavior-preserving.
 */

import { COMPANY_TOOL_NAMES, MARKETPLACE_WRITE_TOOL_NAMES } from "./command-vocabulary.js";

/** The two user-facing states. The ONLY permission concept surfaces see. */
export type PermissionPreset = "ask" | "allow";

// Frozen, mirroring the Python `frozenset` (permissions.py:38) so the required
// set cannot be mutated at runtime to weaken the conformance check.
export const REQUIRED_PRESETS: readonly PermissionPreset[] = Object.freeze(["ask", "allow"]) as readonly PermissionPreset[];

/**
 * Titles are HONEST about current behavior: under the max-permission doctrine
 * both presets run at max, so "ask" must NOT claim to prompt/restrict.
 */
export const PRESET_LABELS: Record<PermissionPreset, { label_key: string; title: string }> = {
  ask: { label_key: "perm.preset.ask", title: "Standard (max today)" },
  allow: { label_key: "perm.preset.allow", title: "Allow all actions" },
};

/**
 * Each preset selects one canonical mode. Owner doctrine (2026-06-22): the
 * underlying runtime is a pure EXECUTION ENGINE always handed its MAXIMUM
 * permission. So BOTH presets map to `bypassPermissions` (permissions.py:81).
 */
export const PRESET_TO_MODE: Record<PermissionPreset, string> = {
  ask: "bypassPermissions",
  allow: "bypassPermissions",
};

export class PresetMapError extends Error {}

/** How a single backend realizes one preset (permissions.py:90). */
export interface PresetRealization {
  /** Human-readable description of the backend-native setting this maps to. */
  readonly native: string;
  /** Does a HUMAN actually get prompted per-action under this preset? Today
   *  this is false everywhere — no human ever sees a prompt. */
  readonly interactive: boolean;
  /** i18n key for an honest, surface-rendered explanation. */
  readonly note_key: string;
  /** Does switching the preset actually change this runtime's behavior? */
  readonly preset_driven: boolean;
}

export type PresetMap = Record<PermissionPreset, PresetRealization>;

/**
 * Construct a PresetRealization with the Python dataclass defaults
 * (preset_driven defaults to true, permissions.py:112). Using this factory keeps
 * the field set complete and uniform — TS interfaces alone do not enforce it.
 */
export function presetRealization(args: {
  native: string;
  interactive: boolean;
  note_key: string;
  preset_driven?: boolean;
}): PresetRealization {
  return {
    native: args.native,
    interactive: args.interactive,
    note_key: args.note_key,
    preset_driven: args.preset_driven ?? true,
  };
}

/** Build a PresetMap so the two-state shape is uniform (permissions.py:122). */
export function makePresets(args: { ask: PresetRealization; allow: PresetRealization }): PresetMap {
  return { ask: args.ask, allow: args.allow };
}

/**
 * Validate a backend's declaration. Throws PresetMapError on any gap — this is
 * the enforcement that makes declaring the mapping a hard requirement for every
 * new backend, and that keeps the preset set at exactly two (no third tier).
 */
export function checkPresetMap(backendName: string, presets: PresetMap): void {
  const keys = new Set(Object.keys(presets));
  const required = new Set<string>(REQUIRED_PRESETS);
  const missing = [...required].filter((k) => !keys.has(k));
  const extra = [...keys].filter((k) => !required.has(k));
  if (missing.length > 0 || extra.length > 0) {
    throw new PresetMapError(
      `backend ${JSON.stringify(backendName)} permission presets must declare exactly ` +
        `${[...REQUIRED_PRESETS].sort().join(", ")}; missing=${missing.sort().join(",")} extra=${extra.sort().join(",")}`,
    );
  }
  for (const preset of REQUIRED_PRESETS) {
    const realization = presets[preset];
    if (realization === null || typeof realization !== "object") {
      throw new PresetMapError(`backend ${JSON.stringify(backendName)} preset ${JSON.stringify(preset)} is not a PresetRealization`);
    }
    if (!(realization.native ?? "").trim()) {
      throw new PresetMapError(`backend ${JSON.stringify(backendName)} preset ${JSON.stringify(preset)} has empty 'native'`);
    }
    if (!(realization.note_key ?? "").trim()) {
      throw new PresetMapError(`backend ${JSON.stringify(backendName)} preset ${JSON.stringify(preset)} has empty 'note_key'`);
    }
    // The Python dataclass guarantees these fields exist and are typed; TS
    // interfaces do not, so validate them to keep parity (a missing/garbage
    // flag must not pass and then serialize as undefined).
    if (typeof realization.interactive !== "boolean") {
      throw new PresetMapError(`backend ${JSON.stringify(backendName)} preset ${JSON.stringify(preset)} 'interactive' must be a boolean`);
    }
    if (typeof realization.preset_driven !== "boolean") {
      throw new PresetMapError(`backend ${JSON.stringify(backendName)} preset ${JSON.stringify(preset)} 'preset_driven' must be a boolean`);
    }
  }
}

/**
 * Serialize a PresetMap for surfaces (permissions.py:151). Mirrors the Python
 * dict comprehension: iterate the INPUT map's entries and return a new object
 * of plain copies — preserving the input's keys, never aliasing the inputs.
 */
export function serializePresetMap(presets: PresetMap): Record<string, PresetRealization> {
  const out: Record<string, PresetRealization> = {};
  for (const [preset, realization] of Object.entries(presets)) {
    out[preset] = { ...realization };
  }
  return out;
}

// ---------------------------------------------------------------------------
// B-class (in-process self-owned) posture helper.
// Only `plan` tightens behavior; every other mode preserves today's behavior.
// ---------------------------------------------------------------------------
export type Posture = "readonly" | "workspace" | "full";

export function postureForMode(mode: string | null | undefined): Posture {
  if (mode === "plan") return "readonly";
  if (mode === "bypassPermissions" || mode === "dontAsk") return "full";
  // default / acceptEdits / auto / unknown -> current workspace behavior
  return "workspace";
}

// Base in-process tools, plus the company-management and marketplace WRITE tool
// names — mirroring permissions.py:191, where _MUTATING_TOOLS folds the kernel-
// state mutation vocabularies in by default so the read-only fence covers them
// by construction. The names come from the single command-vocabulary source
// (command-vocabulary.ts), cross-checked against Python to prevent drift.
export const BASE_MUTATING_TOOLS: readonly string[] = ["run_shell", "write_file", "delegate"] as const;
export const READONLY_TOOLS: readonly string[] = ["read_file", "list_files"] as const;

// Module-private and never aliased out: TS `ReadonlySet` is erased at runtime,
// so handing this object to callers would let them `.delete("company_create")`
// and silently weaken the gate. Python uses an immutable `frozenset`
// (permissions.py:191); we approximate it by keeping the set private and only
// ever returning fresh copies.
const MUTATING_TOOLS: Set<string> = new Set<string>([
  ...BASE_MUTATING_TOOLS,
  ...COMPANY_TOOL_NAMES,
  ...MARKETPLACE_WRITE_TOOL_NAMES,
]);

/** A fresh copy of the full mutating tool set denied under a read-only posture.
 *  Returns a copy so the gate's own set can never be mutated by a caller. */
export function mutatingTools(): ReadonlySet<string> {
  return new Set(MUTATING_TOOLS);
}

/**
 * True when the posture forbids this in-process tool (permissions.py:203).
 * Company/marketplace kernel mutations are covered by default. `extraMutatingTools`
 * is a forward-compat seam for tools registered by not-yet-ported domains.
 */
export function postureDeniesTool(
  posture: Posture,
  toolName: string,
  extraMutatingTools: Iterable<string> = [],
): boolean {
  if (posture !== "readonly") return false;
  if (MUTATING_TOOLS.has(toolName)) return true;
  for (const t of extraMutatingTools) if (t === toolName) return true;
  return false;
}

/** Top-level contract block exported to all surfaces (permissions.py:208). */
export function permissionModeContract(): {
  presets: PermissionPreset[];
  labels: typeof PRESET_LABELS;
  preset_to_mode: Record<PermissionPreset, string>;
} {
  return {
    presets: [...REQUIRED_PRESETS],
    labels: PRESET_LABELS,
    preset_to_mode: { ...PRESET_TO_MODE },
  };
}

File: packages/superclaw/src/superclaw/permissions.py
"""Permission mode framework — the single source of truth for SuperClaw's
two user-facing permission presets and how each Agent Runtime backend maps
them onto its native sandbox/approval mode.

Design rule (see docs/permission-mode-framework.md):

* SuperClaw exposes exactly TWO presets to users: ``ask`` and ``allow``.
* SuperClaw does NOT make per-action decisions of its own — it is a pass-through
  that translates a preset into whatever the underlying runtime already supports.
* Every backend MUST declare ``permission_presets()`` so that adding a new
  Agent Runtime forces the author to decide how its two states are realized.
  A registry-wide conformance test (tests/test_permission_presets.py) fails if
  any registered backend omits or under-specifies the mapping.

This module deliberately stays tiny: no decision engine, no action taxonomy,
no approval coordinator. Those were considered and rejected as over-engineering
(see docs/permission-broker-plan.md for the discarded design).
"""

from __future__ import annotations

from dataclasses import asdict, dataclass
from typing import Any, Literal

# Single source of the company-management tool names (PR-F). Imported at module
# top (company_commands imports only ``models``, so no cycle) so the mutating-tool
# taxonomy below is derived from it rather than hand-copied.
from superclaw.company_commands import COMPANY_TOOL_NAMES as _COMPANY_TOOL_NAMES
from superclaw.marketplace_commands import (
    MARKETPLACE_WRITE_TOOL_NAMES as _MARKETPLACE_WRITE_TOOL_NAMES,
)

# ---------------------------------------------------------------------------
# The two user-facing states. This is the ONLY permission concept surfaces see.
# ---------------------------------------------------------------------------
PermissionPreset = Literal["ask", "allow"]

REQUIRED_PRESETS: frozenset[PermissionPreset] = frozenset({"ask", "allow"})

# Titles are HONEST about current behavior: under the max-permission doctrine
# both presets run at max, so "ask" must NOT claim to prompt/restrict (it would
# be surface fraud). The two slots are kept for a future gated "ask".
PRESET_LABELS: dict[PermissionPreset, dict[str, str]] = {
    "ask": {"label_key": "perm.preset.ask", "title": "Standard (max today)"},
    "allow": {"label_key": "perm.preset.allow", "title": "Allow all actions"},
}

# Each preset selects one canonical PermissionPolicy.mode. The 6 legacy modes
# remain as advanced/back-compat values underneath the two-state shell.
#
# Doctrine (owner decision 2026-06-22): the underlying runtime is a pure
# EXECUTION ENGINE and is always handed its MAXIMUM permission. Governance does
# NOT live in the runtime's own permission system — for the CLI-class backends
# SuperClaw cannot even intercept the runtime's tool loop, and a runtime's own
# headless permission mode (e.g. claude ``acceptEdits``) silently auto-DENIES
# tools that need approval (WebSearch, MCP tools) instead of escalating, which
# surfaced as mysterious tool "Error"s with no way to grant. So BOTH presets map
# to ``bypassPermissions`` — the universal "max" trigger every backend keys on.
#
# HONEST consequence — be explicit, do not overclaim a governance layer that
# does not exist yet: under this mapping the B-class in-process escalation_gate
# (which only fires when posture != "full") is BYPASSED for preset-driven runs,
# and a standard chat/run executes at full posture with NO per-action human gate.
# There is no live human-approval surface today; "governance moves up" is FUTURE
# work (Web approval inbox), not a currently-present door.
#
# What DOES still gate, independent of this mapping (so it is not a TOTAL blanket
# fail-open — these are the only live brakes):
#   * Low-trust / untrusted runs: ContainmentPolicy.permission_mode_floor takes
#     the STRICTEST of (run mode, floor) — an untrusted run stays read-only even
#     under ``bypassPermissions`` (containment.containment_denies_tool). This
#     covers B-class in-process tools; the orchestrator additionally REFUSES
#     backends that cannot prove containment (backend_supports_containment) for
#     low-trust runs, so CLI-class backends never run untrusted code at max.
#   * Explicit ``plan`` mode (advanced) -> readonly posture for review runs.
#   * Pay-switch / scan-intent hard gates live in fusion/plugins and are
#     independent of the permission preset entirely.
#
# Threat model: single user, user-triggered == allowed; only irreversible actions
# warrant confirmation, and that confirmation is deferred to the future human-gate.
PRESET_TO_MODE: dict[PermissionPreset, str] = {
    "ask": "bypassPermissions",
    "allow": "bypassPermissions",
}


# ---------------------------------------------------------------------------
# What a backend declares for one preset.
# ---------------------------------------------------------------------------
@dataclass(frozen=True)
class PresetRealization:
    """How a single backend realizes one preset.

    native:        free-form but human-readable description of the backend-native
                   setting this preset maps to (flags, sandbox+approval, posture).
    interactive:   does a HUMAN actually get prompted per-action at runtime under
                   this preset? Today this is False everywhere: even the Codex
                   app-server's approval callbacks are auto-answered by SuperClaw
                   policy, so no human ever sees a prompt. It may only become True
                   once an approval queue surfaces requests to a person.
    note_key:      i18n key for an honest, surface-rendered explanation (especially
                   required when interactive is False, so users are not misled).
    preset_driven: does switching the preset actually change this runtime's
                   behavior? Declare False explicitly when it cannot (no native
                   gate, or the flag set is fixed) — the conformance test requires
                   preset-driven pairs to differ and no-gate backends to say so.
    """

    native: str
    interactive: bool
    note_key: str
    preset_driven: bool = True

    def to_dict(self) -> dict[str, Any]:
        return asdict(self)


# preset -> PresetRealization
PresetMap = dict[PermissionPreset, PresetRealization]


def make_presets(*, ask: PresetRealization, allow: PresetRealization) -> PresetMap:
    """Build a PresetMap. Backends use this so the two-state shape is uniform."""
    return {"ask": ask, "allow": allow}


def check_preset_map(backend_name: str, presets: PresetMap) -> None:
    """Validate a backend's declaration. Raises ValueError on any gap.

    Used by the registry-wide conformance test AND can be called at registry
    build time. This is the enforcement mechanism that makes declaring the
    mapping a hard requirement for every new backend.
    """
    keys = frozenset(presets)
    if keys != REQUIRED_PRESETS:
        missing = REQUIRED_PRESETS - keys
        extra = keys - REQUIRED_PRESETS
        raise ValueError(
            f"backend {backend_name!r} permission_presets() must declare exactly "
            f"{sorted(REQUIRED_PRESETS)}; missing={sorted(missing)} extra={sorted(extra)}"
        )
    for preset, realization in presets.items():
        if not isinstance(realization, PresetRealization):
            raise ValueError(f"backend {backend_name!r} preset {preset!r} is not a PresetRealization")
        if not (realization.native or "").strip():
            raise ValueError(f"backend {backend_name!r} preset {preset!r} has empty 'native'")
        if not (realization.note_key or "").strip():
            raise ValueError(f"backend {backend_name!r} preset {preset!r} has empty 'note_key'")


def serialize_preset_map(presets: PresetMap) -> dict[str, dict[str, Any]]:
    """Serialize a PresetMap for ui_contracts / surfaces."""
    return {preset: realization.to_dict() for preset, realization in presets.items()}


# ---------------------------------------------------------------------------
# B-class (in-process self-owned) posture helper.
#
# B-class backends (anthropic-agent, gemini-agent) ARE the runtime — there is no
# underlying runtime to delegate to — so SuperClaw enforces a posture directly in
# the tool loop. The posture is derived from the policy mode. Note: only `plan`
# tightens behavior; every other mode preserves today's behavior (zero change).
# ---------------------------------------------------------------------------
Posture = Literal["readonly", "workspace", "full"]


def posture_for_mode(mode: str | None) -> Posture:
    if mode == "plan":
        return "readonly"
    if mode in {"bypassPermissions", "dontAsk"}:
        return "full"
    # default / acceptEdits / auto / unknown -> current workspace behavior
    return "workspace"


# Every in-process (B-class) tool must be classified in exactly one of these
# sets. A conformance test introspects _exec_tool and fails when a new tool is
# added without classifying it here — otherwise read-only posture could be
# silently bypassed by an unclassified mutating tool.
# The company-management tools (PR-F) are KERNEL-STATE mutations (create/update/
# archive a company, hire/update an agent, create/assign/delegate an issue). They
# do not touch the repo filesystem or shell, but they ARE mutating governance
# actions and MUST be denied under a read-only (plan) posture and a low-trust
# containment fence — an untrusted review run must not be able to hire a
# high-privilege agent or restructure a company to escape its sandbox. Folding
# them into _MUTATING_TOOLS makes the SAME fence (posture_denies_tool /
# containment_denies_tool) that already guards run_shell/write_file cover them by
# construction. ``_COMPANY_TOOL_NAMES`` (imported at module top from the single
# command-vocabulary source) ensures this can never drift from what the backend
# actually dispatches.
_MUTATING_TOOLS: frozenset[str] = frozenset(
    {"run_shell", "write_file", "delegate"}
    | set(_COMPANY_TOOL_NAMES)
    # Marketplace WRITES (post/bid/claim/submit/abandon/accept/accept_bid) cross an
    # external boundary + make commitments — mutating by construction, so the same
    # read-only-posture / low-trust-containment fence that guards run_shell covers
    # them too. Reads (browse/inspect) are NOT here (they do not mutate).
    | set(_MARKETPLACE_WRITE_TOOL_NAMES)
)
_READONLY_TOOLS: frozenset[str] = frozenset({"read_file", "list_files"})


def posture_denies_tool(posture: Posture, tool_name: str) -> bool:
    """True when the posture forbids this in-process tool."""
    return posture == "readonly" and tool_name in _MUTATING_TOOLS


def permission_mode_contract() -> dict[str, Any]:
    """Top-level contract block exported to all surfaces via ui_contracts."""
    return {
        "presets": list(REQUIRED_PRESETS),
        "labels": PRESET_LABELS,
        "preset_to_mode": dict(PRESET_TO_MODE),
    }
